os/pm/pm_idle.c

```diff
@@ -65,12 +69,45 @@ void pm_idle(void)
 	enum pm_state_e newstate;
 #ifdef CONFIG_PM_TIMEDWAKEUP
 	clock_t delay;
+#endif
+#ifdef CONFIG_SMP
+	int me = this_cpu();
+	int cpu;
+	int paused_count = 0;
+	FAR struct tcb_s *tcb;
 #endif
 	/* Decide, which power saving level can be obtained */
 	flags = enter_critical_section();
 	newstate = pm_checkstate();
 	/* Perform state-dependent logic here */
-	pmvdbg("newstate= %d\n", newstate);
+	/* For SMP case, we need to check secondary core status
+	   If secondary core status is not in idle thread, abort
+	   the sleep and check again on next cycle
+	*/
+#ifdef CONFIG_SMP
+	if (newstate == PM_SLEEP) {
+		for (cpu = 1; cpu < CONFIG_SMP_NCPUS; cpu++) {
+			/* If the CPU is just back from sleep, abort the sleep */
+			if (up_get_secondary_cpu_state(cpu) == CPU_WAKE_FROM_SLEEP) {
+				goto EXIT;
+			}
+			/* Gate the cpu first, before checking which task it is handling */
+			up_cpu_gating(cpu);
+			/* Polling the status of the target CPU, 
+			   The flag should be > 0 if it is already gated
+			*/
+			while(!up_get_gating_flag_status(cpu));
+			tcb = current_task(cpu);
+			/* Check if current cpu is in idle thread, and whether there is pending
+			   pause request on primary core 
+			*/
+			if (tcb->pid != cpu || up_cpu_pausereq(me)) {
```

Comment:
do wa need to check still up_cpu_pausereq?

Reply:
I think we could remove it.